Add missing Time Evolution Tutorials

christian512 opened the pull request:
This adds more tutorials on steady state and stochastic solver.

hodgestar commented on tutorials-v4/measures-trajectories-cats-kerr.md:
```suggestion
plt.plot(tlist, sol_mc_mean.expect[0], "r", label="Counting")
```

hodgestar reviewed (changes requested):
I ended up doing a reasonably thorough review because I got interested in the optomechanical example after this morning's discussion. :)

christian512 commented on tutorials-v4/optomechanical-steadystate.md:
````diff
@@ -0,0 +1,203 @@
+It is known that the density matrix for the mechanical oscillator is diagoinal in the Fock basis due to phase diffusion.  However some small off-diagonal terms show up during the factorization process
+
+```python
+hinton(rho_mech, xlabels=[""] * Nm, ylabels=[""] * Nm);
````
I agree. I added a `plt.spy()` plot to show that there exist off-diagonal terms.

christian512 commented:
Thanks for the review! :) I updated the tutorial accordingly. 

I also reduced the number of solvers used in `optomechanical-steadystate` to have a more reasonable runtime.

nwlambert commented on tutorials-v4/optomechanical-steadystate.md:
```diff
@@ -0,0 +1,213 @@
+kappa = 0.3  # Cavity damping rate
+E = 0.1  # Driving Amplitude
+g0 = 2.4 * kappa  # Coupling strength
+Qm = 1e4  # Mech quality factor
```
just a minor thing, looking at the new hinton plots with this reduced Nm cut-off technically speaking the results might suffer from truncation errors. I think for the purpose of the notebook this is not particularly important, but could be mitigated anyway by reducing Qm by a factor of 10 or so.

christian512 replied:
I'm not sure what you mean by the "reduced Nm cut-off". However, in the notebook we can not see the truncation errors, right? I would leave this as is for now.

nwlambert replied:
yeah, sorry, to clarify I ran the same parameters with the original value of Nm=80 (I think it was?) and the new value of Nm=30 and looking at the hinton plots, and the Nm=80 has significant diagonal components for fock states >30, so its just that the model intrinsically has a truncation error for Nm=30.   To make it run quicker and suppress that error you can just ramp up the mechanical damping, which corresponds to reducing the mechanical quality factor Qm.   But yeah, unless the user explicitly run the notebook by changing Nm this error isn't apparent.

christian512 replied:
I have reduced the quality factor slightly, such that there are vanishing diagonal components for `Nm > 30` and that the `hinton` plot is still somewhat interesting (not only first two diagonals).

christian512 commented on tutorials-v4/quasi-steadystate-driven-system.md:
```diff
@@ -0,0 +1,125 @@
+    U = propagator(hamiltonian_t, T, c_op_list, H_args)
+
+    rho_ss = propagator_steadystate(U)
+
```
Yes, I agree. Did so in the last commit.

christian512 commented:
I have moved the tutorials into the `time-evolution` folder, which is required later on for creating the index page.I also updated the titles of the notebooks to match the titles at the [tutorials website](https://qutip.org/tutorials.html).

Ericgig commented on tutorials-v5/time-evolution/010_brmesolve_phonon_interaction.md:
```diff
@@ -213,8 +213,8 @@ def brme_step(args):
-# use QuTiP's builtin parallelized for loop, parfor
-results = parfor(brme_step, itertools.product(wd_list, Om_list))
+# use QuTiP's builtin parallelized for loop: parallel_map
+results = parallel_map(brme_step, list(itertools.product(wd_list, Om_list)))
```
It would be nice if this example should be using the class interface instead of the function call in v5.
The class interface is to be used when computing the same system with different parameters, which is what is done here.

christian512 replied:
Thanks for the review @Ericgig! I changed from using `brmesolve` to `BRSolver` in the last commit. `BRSolver` currently requires that the spectra for the `a_ops` are given as `coefficient`. Hence, I can not pass a string spectrum.

I think it would be great if `BRSolver` also accepts other types of spectra.

hodgestar commented on tutorials-v5/time-evolution/010_brmesolve_phonon_interaction.md:
````diff
@@ -259,3 +256,7 @@ assert inv_mat_X[2][-1] > 0.9
 # for -1meV detuning steadystate close to ground state
 assert inv_mat_X[0][-1] < 0.1
 ```
+
+```python
+
+```
````
Empty python block to remove.

hodgestar commented on tutorials-v4/time-evolution/019_optomechanical-steadystate.md:
````diff
@@ -0,0 +1,217 @@
+```python
+about()
+```
+
+```python
+
+```
````
I guess we could remove this empty python block too.

hodgestar reviewed (commented):
Also looks good. Some questions:

- Are notebooks 15, 16 and 17 in a separate PR?
- It would be nice to automatically trim empty python cells at the end of the notebooks, or at least to warn about them in the test runs, but maybe just create an issue for that?

christian512 commented:
Thanks for the review, @hodgestar! 
The notebooks 15, 16, 17 are the Stochastic Solver notebooks in #30. 

I opened an issue to create a test for empty code cells in #38.
